File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/encryption/secrets/UserLocalEncryptionSecretSpec.kt
package stasis.test.client_android.lib.encryption.secrets

import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.ByteString.Companion.decodeBase64
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.encryption.secrets.UserLocalEncryptionSecret

class UserLocalEncryptionSecretSpec : WordSpec({
    "A UserLocalEncryptionSecret" should {
        val encryptionIv = "J9vRvveXTnC0iF4ymYbIo5racLWx60CGxcOlklH/qH4xqIKvlsZQyr66bGFxzpYrayRS7iipCVimlYt7BCj7uQ=="
        val encryptionKey = "nXT1Bw0YCrk79xgnvlUJ5CZByYD9nuSZo9XQghf1xQU="
        val secret = "BOunLSLKxVbluhDSPZ/wWw=="

        val localEncryptionSecret = UserLocalEncryptionSecret(
            user = SecretsConfig.testUser,
            iv = encryptionIv.decodeBase64()!!,
            key = encryptionKey.decodeBase64()!!,
            target = SecretsConfig.testConfig
        )

        val deviceSecret = DeviceSecret(
            user = SecretsConfig.testUser,
            device = SecretsConfig.testDevice,
            secret = secret.decodeBase64()!!,
            target = SecretsConfig.testConfig
        )

        val encryptedDeviceSecret = "z+pus9Glc/HssFVWozd+T2iuw4OOM4aeqcdDY+gvG8M=".decodeBase64()!!

        "encrypt device secrets" {
            val actualEncryptedSecret = localEncryptionSecret.encryptDeviceSecret(
                secret = deviceSecret
            )

            actualEncryptedSecret shouldBe (encryptedDeviceSecret)
        }

        "decrypt device secrets" {
            val actualDeviceSecret = localEncryptionSecret.decryptDeviceSecret(
                device = SecretsConfig.testDevice,
                encryptedSecret = encryptedDeviceSecret
            )

            actualDeviceSecret shouldBe (deviceSecret)
        }

        "not render its content via toString" {
            localEncryptionSecret.toString() shouldBe ("Secret(${localEncryptionSecret.javaClass.name})")
        }
    }
})

File: client/src/test/scala/stasis/test/specs/unit/client/encryption/secrets/UserLocalEncryptionSecretSpec.scala
package stasis.test.specs.unit.client.encryption.secrets

import org.apache.pekko.actor.ActorSystem

import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.encryption.secrets.UserLocalEncryptionSecret
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.EncodingHelpers

class UserLocalEncryptionSecretSpec extends AsyncUnitSpec with SecretsConfig with EncodingHelpers {
  "A UserLocalEncryptionSecret" should "encrypt device secrets" in {
    localEncryptionSecret
      .encryptDeviceSecret(
        secret = deviceSecret
      )
      .map { actualEncryptedSecret =>
        actualEncryptedSecret should be(encryptedDeviceSecret)
      }
  }

  it should "decrypt device secrets" in {
    localEncryptionSecret
      .decryptDeviceSecret(
        device = testDevice,
        encryptedSecret = encryptedDeviceSecret
      )
      .map { actualDeviceSecret =>
        actualDeviceSecret should be(deviceSecret)
      }
  }

  it should "not render its content via toString" in {
    localEncryptionSecret.toString should be(s"Secret(${localEncryptionSecret.getClass.getName})")
  }

  private implicit val system: ActorSystem = ActorSystem(name = "UserEncryptionSecretSpec")

  private val encryptionIv = "J9vRvveXTnC0iF4ymYbIo5racLWx60CGxcOlklH/qH4xqIKvlsZQyr66bGFxzpYrayRS7iipCVimlYt7BCj7uQ=="
  private val encryptionKey = "nXT1Bw0YCrk79xgnvlUJ5CZByYD9nuSZo9XQghf1xQU="
  private val secret = "BOunLSLKxVbluhDSPZ/wWw=="

  private val localEncryptionSecret = UserLocalEncryptionSecret(
    user = testUser,
    iv = encryptionIv.decodeFromBase64,
    key = encryptionKey.decodeFromBase64
  )

  private val deviceSecret = DeviceSecret(
    user = testUser,
    device = testDevice,
    secret = secret.decodeFromBase64
  )

  private val encryptedDeviceSecret = "z+pus9Glc/HssFVWozd+T2iuw4OOM4aeqcdDY+gvG8M=".decodeFromBase64
}
